contract_manager/src/contracts/ton.ts

```diff
@@ -0,0 +1,286 @@
+  }
+
+  async getPriceFeed(feedId: string): Promise<PriceFeed | undefined> {
+    const contract = await this.getContract();
+    const feedIdWithPrefix = `0x${feedId}`;
+    try {
+      const price = await contract.getPriceUnsafe(feedIdWithPrefix);
+      const emaPrice = await contract.getEmaPriceUnsafe(feedIdWithPrefix);
+      return {
+        price: {
+          price: price.price.toString(),
+          conf: price.conf.toString(),
+          expo: price.expo.toString(),
+          publishTime: price.publishTime.toString(),
+        },
+        emaPrice: {
+          price: emaPrice.price.toString(),
+          conf: emaPrice.conf.toString(),
+          expo: emaPrice.expo.toString(),
+          publishTime: emaPrice.publishTime.toString(),
+        },
+      };
+    } catch (e) {
+      console.error(e);
+      return undefined;
+    }
+  }
+
+  async getValidTimePeriod(): Promise<number> {
+    // Not supported but return 1 because it's required by the abstract class
+    return 1;
+  }
+
+  async getWormholeContract(): Promise<TonWormholeContract> {
+    // Price feed contract and wormhole contract live at same address in TON
+    return new TonWormholeContract(this.chain, this.address);
+  }
+
+  async getBaseUpdateFee() {
+    const contract = await this.getContract();
+    const amount = await contract.getSingleUpdateFee();
+    return {
+      amount: amount.toString(),
+      denom: this.chain.getNativeToken(),
+    };
+  }
+
+  async getDataSources(): Promise<DataSource[]> {
+    const contract = await this.getContract();
+    const dataSources = await contract.getDataSources();
+    return dataSources.map((ds: DataSource) => ({
+      emitterChain: ds.emitterChain,
+      emitterAddress: ds.emitterAddress.replace("0x", ""),
+    }));
+  }
+
+  async getGovernanceDataSource(): Promise<DataSource> {
+    const contract = await this.getContract();
+    const result = await contract.getGovernanceDataSource();
+    if (result === null) {
+      throw new Error("Governance data source not found");
+    }
+    return {
+      emitterChain: result.emitterChain,
+      emitterAddress: result.emitterAddress,
+    };
+  }
+
+  async executeUpdatePriceFeed(
+    senderPrivateKey: PrivateKey,
+    vaas: Buffer[]
+  ): Promise<TxResult> {
+    const client = await this.chain.getClient();
+    const contract = await this.getContract();
+    const wallet = await this.chain.getWallet(senderPrivateKey);
+    const sender = await this.chain.getSender(senderPrivateKey);
+    for (const vaa of vaas) {
+      const fee = await contract.getUpdateFee(vaa);
+      console.log(fee);
+      await contract.sendUpdatePriceFeeds(
+        sender,
+        vaa,
+        156000000n + BigInt(fee)
+      ); // 156000000 = 390000 (estimated gas used for the transaction, this is defined in target_chains/ton/contracts/common/gas.fc as UPDATE_PRICE_FEEDS_GAS) * 400 (current settings in basechain are as follows: 1 unit of gas costs 400 nanotons)
```

Comment:
do we want to add it to the sdk for people so they use it without hardcoding?

Reply:
yeah good point